library/platform_util.c

```diff
@@ -165,13 +165,37 @@ uint32_t mbedtls_platform_random_in_range( size_t num )
 #endif
 }
 
-/* Some compilers (armcc 5 for example) optimize away successive reads from a
- * volatile local variable (which we use as a counter-measure to fault
- * injection attacks), unless there is a call to an external function between
- * them. This functions doesn't need to do anything, it just needs to be
- * in another compilation unit. So here's a function that does nothing. */
-void mbedtls_platform_enforce_volatile_reads( void )
+int mbedtls_platform_random_delay( size_t max_rand )
 {
+#if !defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
+    (void) max_rand;
+    return -1;
+#else
+    size_t rn_1, rn_2, rn_3;
+    volatile size_t i = 0;
+    uint8_t shift;
+    if( max_rand == 0 || max_rand > INT_MAX )
```

Comment:
Major - I think this check is vulnerable to fault injection attack. You can force an early return of the function.

The problem here - is if you do a first fault injection, this check will be a *fixed* period after the first, and allow you to bypass the secondary check.

Sorry, but this is flawed.

Reply:
One thing to note is that this is used between double checks in other code, so potential attacker would need 3 consecutive faults to be injected to possibly benefit from skipping the random delay. Anyway, this check would probably more preferably just use some default value for the max_rand instead of returning -1. (As the return values aren't checked anywhere)